src/bosh-director/lib/bosh/director/jobs/scheduled_tasks_cleanup.rb

```diff
@@ -15,6 +15,27 @@ def initialize(_param = {})
         @task_remover = Bosh::Director::Api::TaskRemover.new(Config.max_tasks)
       end
 
+      def update_orphaned_tasks_with_state_error(result)
+        actual_delayed_job_scheduled_task_ids = []
+        Delayed::Worker.backend = :sequel
+        Delayed::Job.all.select do |job|
+          actual_delayed_job_scheduled_task_ids << YAML.safe_load(job.handler)['task_id']
+        end
+
+        errored_tasks = []
+        Bosh::Director::Models::Task.select.where('state': %w[queued processing]).each do |task|
+          next if actual_delayed_job_scheduled_task_ids.include?(task.id)
+
+          errored_tasks << task.id
+          task.state = 'error'
+          task.save
+        end
```

Comment:
This is tricky, and I'm not sure how to solve it.

You select the delayed jobs first from the database, then select the tasks. If a task doesn't exist in the delayed job records, it gets errored.

But what if a task is created AFTER you pull the delayed jobs records from the database? Then a new task is going to get marked as errored incorrectly.

I'm not sure you can reverse the order of the finds either, you have the same problem where either record could be modified. Doing this work in a transaction and setting the isolation level to `Repeatable read` might work, but it is a bit risky and I've never done this before. 

https://www.postgresql.org/docs/current/transaction-iso.html

You'd have to ensure the isolation level gets reset at the end, not sure if Sequel does that or not, it might keep it for the connection if you don't turn it back down. And then you have the problem of the database raising errors if there is a serialization failure where the Task record is updated in a different transaction at the same time. This might be fine if the error only happens here, we could rescue it and retry the function. But if the error can happen on a normal transaction elsewhere that wouldn't be great.

Maybe there is some other way of doing this where you collect up all the potential errored tasks, then double check at the end that there are no delayed jobs for them, and then when you go to update them to `error`, you'd need to make sure they still exist in the database? Not 100% sure that will catch all the edge cases, but it might...

Reply:
So I think it should be safe because:

A task gets created by something calling JobQueue.enqueue

JobQueue.enqueue first creates the task by calling 
create_task. Within create_task we call task.save. At this point the task exists in the bosh tasks db, finally create_task returns the task object

then we create the db_job (using, among other things the value of task.id ) 
and use the db_job object and pass it into DelayedJob.enqueue

At this point the delayed job exists in the delayed job db.. 

So we cannot have `but what if a task is created AFTER you pull the delayed jobs records from the database? Then a new task is going to get marked as errored incorrectly.` being the case

Reply:
how about using the timestamp? 

If the task got updated / created within the last day, the cleanup is skipped

Reply:
Probably just a couple of minutes is fine actually. But yeah, that seems like a good solution. `then` isn't normally used in ruby, and there is an extra `)` in there too. I would also expect the ORM to return the `timestamp` as a DateTime object already so the `parse` isn't probably needed.

I was kind of surprised that `-` on a `DateTime` object just flows up to the `Date` class and gives you the number of days difference. If you want smaller offset you can `to_time` the object first in which case `-` will give you a seconds offset.

Reply:
Thanks. I updated it to five minutes.. Also updated the test to check for that scenario